### ISSUE TITLE
Density scaling and bandwidth in type_ridge()

### ISSUE THREAD

**Issue by zeileis** [grantmcdermott/tinyplot#271]
In `get_density()` within `type_ridge()` the density estimates are scaled to the same maximum which I believe is incorrect. Also the bandwidths are by default computed separately whereas `ggridges` employs a joint bandwidth estimate which typically seems to be better. But maybe I'm overlooking something here, Vincent @vincentarelbundock ?

For illustration let's consider a sample with two groups with very different variances:

```
set.seed(0)
d <- data.frame(
  y = rep(1, 100),
  x = c(rnorm(50), rnorm(50, mean = 3, sd = 0.2)),
  z = factor(rep(1:2, each = 50))
)
```

Using `ggridges` a joint bandwith of 0.218 is selected and the resulting densities have very different maxima - as you would expect if both should have the area 1.

```
ggplot(d, aes(x = x, y = z)) + geom_density_ridges() + theme_minimal()
ggplot(d, aes(x = x, y = y, group = z)) + geom_density_ridges() + theme_minimal()
```

![tinyplot-ridge1](https://github.com/user-attachments/assets/7d848cc2-776f-4b59-8e7b-681948046b1e)

In contrast, because `get_density()` scales the density to have the same maximum, the second group has a much smaller area than the first in `tinyplot(..., type = "ridge")`.

```
tinyplot(z ~ x, data = d, type = "ridge", grid = TRUE)
tinyplot(y ~ x | z, data = d, type = "ridge", grid = TRUE)
```

![tinyplot-ridge2](https://github.com/user-attachments/assets/36fcf76b-e505-4cac-be38-f0d04630fff7)

However, if I drop the scaling in line https://github.com/grantmcdermott/tinyplot/blob/main/R/type_ridge.R#L99 and select the same bandwidth of 0.218, then I get virtually the same result as in `ggridges`.

```
tinyplot(z ~ x, data = d, type = type_ridge(bw = 0.218), grid = TRUE)
tinyplot(y ~ x | z, data = d, type = type_ridge(bw = 0.218), grid = TRUE)
```

![tinyplot-ridge3](https://github.com/user-attachments/assets/5a08c422-4202-4dcb-a172-01d65f6f464e)

Note that `type = "density"` agrees here and chooses the same joint bandwidth without applying any scaling of the maximum:

```
tinyplot(~ x | z, data = d, type = "density", grid = TRUE)
```

![tinyplot-ridge4](https://github.com/user-attachments/assets/988afded-822d-4e36-bcdc-f06f085a92cd)

**Comment by vincentarelbundock** [comment]
Oh that's interesting. I chose this purely for visual reason, and had no principled reason. Sorry!

I'll defer to you on best defaults.

**Comment by zeileis** [comment]
OK, good, thanks for the quick feedback. Removing the scaling to the same maximum is straightforward.

Regarding the bandwidth: Both `ggridges` and `type_density()` seem to use the average of the individual bandwidths per group, see: https://github.com/wilkelab/ggridges/blob/master/R/stats.R#L109-L112 and https://github.com/grantmcdermott/tinyplot/blob/main/R/type_density.R#L83-L86

As the code is so similar: Is the `type_density` code inspired by `ggridges` or are they both inspired by something else? Why is this simply using the mean rather than the weighted mean (so that larger groups would receive more weight)?

Should we then always enforce that the same bandwith is used throughout? Or should we allow different bandwidth per ridgeline? If the latter, how should we specify this, with an additional argument or does anyone have a better idea?

**Comment by zeileis** [comment]
OK, thanks for the explanation. In any case, I think that `type_density` and `type_ridge` should use the same default bandwidth. And I lean towards `weighted.mean` but I haven't explored this more systematically to check whether it really works better in unbalanced data.

**Comment by grantmcdermott** [comment]
Note: with #284 we ultimately opted with individual bandwidths in (regular) grouped density plots... albeit with the possibility for users to override via the `joint.bw` argument.

Should we update the `type_ridge` code to do so as well?

**Comment by zeileis** [comment]
Yes, `type_density()` and `type_ridge()` should be consistent in their handling of the bandwidths, I think.

I'll try to have a closer look at the details in `type_density()` later tonight and then follow-up again.

**Comment by grantmcdermott** [comment]
I was looking at this again today. One discrepancy from the `ggridges` side is that it treats joint bandwidth selection differently across groups and facets. Specifically, a common bandwidth is used across groups but distinct bandwidths are used across facets (although every group within a facet will receive the same bandwidth). E.g., using Achim's running example from above:

```r
ggplot(d, aes(x = x, y = y, group = z)) + geom_density_ridges() 
#> Picking joint bandwidth of 0.218
ggplot(d, aes(x = x, y = y)) + geom_density_ridges() + facet_wrap(~ z)
#>Picking joint bandwidth of 0.377
#> Picking joint bandwidth of 0.0591
```

I don't really get why you'd want to treat facets and groups differently here. At the same time, I also think that some kind of joint bandwidth makes sense as a default for ridge plots (contra our decision in #284  to go with individual bandwidths as the default for `type_density()`).

My current inclination is thus to default to a joint bandwidth based on the `"full"` dataset as a default. We can offer `"owm"` (observation-weighted mean) as an alternative joint bandwidth, as well as `"none"` if users want to opt out of joint bandwidths entirely and calculate the bandwidth for each distribution on its own.

P.S. I'm pretty sure `ggridges` is using a simple mean to calculate its joint bandwidth, not a weighted mean or full sample bandwidth as we discussed.

**Comment by grantmcdermott** [comment]
Okay, here's an example of my proposed changes in action. If you are both happy, then I'll role these in to #270. (Which we can hopefully then merge.)

```r
pkgload::load_all("~/Documents/Projects/tinyplot/")
tpar(grid = TRUE)

airq = airquality |>
  transform(
    Late = ifelse(Day > 15, "Late", "Early"),
    Month = factor(Month)
  )

tinyplot(
  Month ~ Ozone, facet = ~ Late, data = airq,
  type = type_ridge(), # joint.bw = "owm" (proposed default)
  main = 'joint.bw = "own" (obs.-weighted mean)'
)
```

![Image](https://github.com/user-attachments/assets/afcba9df-b676-4121-b03b-040f62288349)

```r
tinyplot(
  Month ~ Ozone, facet = ~ Late, data = airq,
  type = type_ridge(joint.bw = "full"), 
  main = 'joint.bw = "full"'
)
```

![Image](https://github.com/user-attachments/assets/272e9a49-3bd7-4f44-ad2d-868490076d2b)

```r
tinyplot(
  Month ~ Ozone, facet = ~ Late, data = airq,
  type = type_ridge(joint.bw = "none"), 
  main = 'joint.bw = "none"'
)
```

![Image](https://github.com/user-attachments/assets/337e2fea-78e0-4d0a-b0d1-be453838f99d)

Note: As thing stand right now, the default for #270 would actually be the `"none"` option that is shown last here. I don't think this looks good (even if we decided to make it the default for "vanilla" `type_density()`).
